### PR TITLE
[Feature] Add Custom Authentication method

### LINKED ISSUE #43540: [Feature] CUSTOM authentication support
## Feature request

**Is your feature request related to a problem? Please describe.**

in company, we have an unify ID and password within any platforms. To use StarRocks better, we hope to use the same authentication method without extra management.

**Describe the solution you'd like**

An interface which can pass username and password to our authentication server.

**Describe alternatives you've considered**

I considered to add an SQL Gateway before StarRocks to check user. However, it's more complex for the whole structure. 
﻿
**Additional context**

### PR DESCRIPTION
**Pull request by lianneli**
## Why I'm doing:

Add a Custom Authentication method for user

## What I'm doing:

Fixes #43540

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by nshangyiming** [comment]
@lianneli Thanks for your contribution, we will give feedback later.

File: fe/fe-core/src/main/java/com/starrocks/authentication/CustomAuthenticationProvider.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+// limitations under the License.
+
+package com.starrocks.authentication;
+
+import com.starrocks.mysql.MysqlPassword;
+import com.starrocks.mysql.privilege.AuthPlugin;
+import com.starrocks.mysql.privilege.Password;
+import com.starrocks.sql.ast.UserIdentity;
+
+public abstract class CustomAuthenticationProvider implements AuthenticationProvider {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/authentication/CustomAuthenticationProvider.java:22]: Hi @lianneli , i understand what you want to do, but seems that you don't need to add an extra `CustomAuthenticationProvider` class, `AuthenticationProvider` is designed to achieve that.

Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/authentication/CustomAuthenticationProvider.java:22]: If you want to add an extra authentication method, you can refer to `KerberosAuthenticationProvider`. It uses a standalone jar file to enable kerberos authentication for Sr.

Comment (lianneli) [reply fe/fe-core/src/main/java/com/starrocks/authentication/CustomAuthenticationProvider.java]: Thanks for your advice! I delete `CustomAuthenticationProvider` and add the Custom implementation for check password methods in `Password` class. If it's ok?

Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/authentication/CustomAuthenticationProvider.java]: @lianneli how do you specify the server address where you auth with when calling `customAuthenticationProvider.authenticate(remoteUser, "", remotePassword, randomString, null);`?

Comment (lianneli) [reply fe/fe-core/src/main/java/com/starrocks/authentication/CustomAuthenticationProvider.java]: For username which created and saved in StarRocks, the server address and other params should be setted in user's jar, for which StarRocks do not need to consider.
﻿
As for other usernames, I find that you have designed `create security integration` SQL syntax which can transfer params to Custom Class. Users can set config `authentication_chain` properly and get those params by the code: 
```java
CustomSecurityIntegration securityIntegration = (CustomSecurityIntegration) authenticationInfo.extraInfo.get("SECURITY_INTEGRATION"); 

Map<String, String> userParams = securityIntegration.getPropertyMap();

```
﻿
by the way, I want to complete the implementation about `Drop security integration` and `Alter security integration` recently, to make the second scenario more convenient to use.

Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/authentication/CustomAuthenticationProvider.java]: @lianneli SecurityIntegration should be used with Config.authentication_chain, and acutally SecurityIntegration is an enterprise feature. For some reason, we left some of the code in community edition, and these remain code will probably be removed in the future, so you can't rely on SecurityIntegration to implement what you want to do. Considering this situation, i think the best choice is to refer to `KerberosAuthenticationProvider` and maintain your own patch for now. Sorry for your time and we still hope for your contribution in the future.

Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/authentication/CustomAuthenticationProvider.java]: Oh, btw, if you have any interest of our enterprise edition, i can ask our pre-sales engineer to contact you.

Comment (lianneli) [reply fe/fe-core/src/main/java/com/starrocks/authentication/CustomAuthenticationProvider.java]: @nshangyiming It's really regrettable that you donot provide security integration related SQL syntax! I can only maintain my patch for company's scenario. 
Custom authentication for users in SR is OK?

Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/authentication/CustomAuthenticationProvider.java]: Yes, if you don't rely on `SecurityIntegration`, you can still consider how to implement a custom authenticator, and that is absolutely welcome. BTW, i think trino already has a good design: https://trino.io/docs/current/develop/password-authenticator.html , you can refer to that as a start point and consider how to integrate it with SR. Thanks again.

Comment (lianneli) [reply fe/fe-core/src/main/java/com/starrocks/authentication/CustomAuthenticationProvider.java]: @nshangyiming Thanks for your advice. I have changed the implementation with independent conf file, and do not dependency security integration. Could you please review again?

**Pull request closed without merging**